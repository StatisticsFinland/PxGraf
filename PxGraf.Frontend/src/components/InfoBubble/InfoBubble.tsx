import { Popper } from "@mui/material";
import React from "react";
import styled from "styled-components";
import InfoIcon from '@mui/icons-material/Info';
import { useTranslation } from "react-i18next";
interface IInfoBubbleProps {
    info: React.ReactNode;
    ariaLabel: string;
    placement?: 'auto' | 'auto-end' | 'auto-start' | 'top' | 'top-end' | 'top-start' | 'left' | 'left-end' | 'left-start' | 'right' | 'right-end' | 'right-start' | 'bottom' | 'bottom-end' | 'bottom-start';
    id?: string;
}

const InfoButton = styled.button`
    background: transparent;
    border: none;
`;

const PopperInfo = styled.div`
    background-color: #0073b0;
    color: #fff;
    border: none;
    padding: 16px;
    box-shadow: 3px 3px 10px #737373;
    border-radius: 5px;
    max-width: 300px;
`;

const StyledPopper = styled(Popper)`
    z-index: 999;
`;

export const InfoBubble: React.FC<IInfoBubbleProps> = ({ info, ariaLabel, placement = 'auto', id = null }) => {
    const [open, setOpen] = React.useState(false);
    const anchorElement = React.useRef(null);
    const { t } = useTranslation();

    // Event listener for pressing the escape key to close the info bubble
    React.useEffect(() => {
        const handleEscape = (event: KeyboardEvent) => {
            if (event.key === 'Escape') {
                setOpen(false);
            }
        };

        if (open) {
            window.addEventListener('keydown', handleEscape);
        } else {
            window.removeEventListener('keydown', handleEscape);
        }

        return () => {
            window.removeEventListener('keydown', handleEscape);
        };
    }, [open]);

    return (
        <>
            <InfoButton
                id={id}
                aria-label={`${open ? t('tooltip.close') : t('tooltip.open')}: ${ariaLabel}`}
                onMouseEnter={() => setOpen(true)}
                onMouseLeave={() => setOpen(false)}
                onClick={() => setOpen(!open)}
                ref={anchorElement}
            >
                <InfoIcon color={'info'} />
            </InfoButton>
            <StyledPopper role="alert" keepMounted popperOptions={{ placement: placement }} open={open} anchorEl={anchorElement.current} onMouseEnter={() => setOpen(true)} onMouseLeave={() => setOpen(false)} >
                <PopperInfo>
                    {info}
                </PopperInfo>
            </StyledPopper>
        </>
    );
}

export default InfoBubble;