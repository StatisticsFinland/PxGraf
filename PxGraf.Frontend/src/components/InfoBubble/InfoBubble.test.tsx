import { fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import InfoBubble from './InfoBubble';

describe('Rendering test', () => {
    it('renders correctly with popper hidden', () => {
        const { asFragment } = render(<InfoBubble info='foobar' ariaLabel='foo' />);
        expect(asFragment()).toMatchSnapshot();
    });

    it('renders correctly with popper visible', () => {
        const { asFragment } = render(<InfoBubble info='foobar' ariaLabel='foo' isOpen={true} />);
        expect(asFragment()).toMatchSnapshot();
    });
});

describe('Assertion test', () => {
    it('renders popper when clicked', () => {
        render(<InfoBubble info='foobar' ariaLabel='foo' />);
        fireEvent.click(screen.getByRole('button'));
        expect(screen.getByText('foobar')).toBeInTheDocument();
    });
    it('renders popper when mouse enters', () => {
        render(<InfoBubble info='foobar' ariaLabel='foo' />);
        fireEvent.mouseEnter(screen.getByRole('button'));
        expect(screen.getByText('foobar')).toBeInTheDocument();
    });
});
