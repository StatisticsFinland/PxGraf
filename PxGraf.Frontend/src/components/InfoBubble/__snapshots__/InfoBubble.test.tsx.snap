// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering test renders correctly with popper hidden 1`] = `
<DocumentFragment>
  <button
    aria-label="tooltip.open: foo"
    class="sc-KXCwU LyzUe"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-colorInfo MuiSvgIcon-fontSizeMedium css-1yq1soz-MuiSvgIcon-root"
      data-testid="InfoIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m1 15h-2v-6h2zm0-8h-2V7h2z"
      />
    </svg>
  </button>
</DocumentFragment>
`;
